### DISCUSSION TITLE
OC 与 C++ 混编导致 UTF8String 野指针崩溃 #2397

### DISCUSSION THREAD

**Discussion by hntTheBee** [Tencent/libpag#2529] [Q&A]
之前遇到过的问题，在issue里提过没有得到回复。
https://github.com/Tencent/libpag/issues/2397
连续调用 PAGImageView 的 - (void)setPathAsync:completionBlock: 出现崩溃
1 当第一次调用 PAGImageView 的 - (void)setPathAsync:completionBlock: 时，下面函数会去下载 PAG 文件:
2 如果此时同一个 PAGImageView 对象触再次触发 - (void)setPathAsync:completionBlock: 调用，[filePath release]会将 PAGImageView 持有的文件 path 释放掉:
第一步的网络下载成功之后，调用后面的 [PAGDiskCacheImpl WritFile:path data:cacheData]; 方法访问了 path 变量，此时这个 path 变量已经被 release 了，会崩在这个方法的内部:
因为下载是在一个 C++ block 里面执行的:但是 C++ block 捕获 path 变量并不会增加 OC 对象的引用计数，导致前面第 2 次调用 - (void)setPathAsync:completionBlock: 释放了 path，访问了野指针崩溃
同时，如果直接调用 PAGFile 的 LoadAsync:completionBlock:方法，这个问题也容易出现，原因和上面是一样的。

**Comment by hntTheBee** [comment]
还有#2396 PAGImageView 同步下载由于持有锁导致的 watchdog 问题 
都已经定位到问题了，希望能修复一下，至少回复一下orz

**Reply by kevingpqi123** [reply]
https://github.com/Tencent/libpag/discussions/2532
后续统一在这个帖子中回复吧
